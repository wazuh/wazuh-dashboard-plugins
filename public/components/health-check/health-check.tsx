import React, { Component } from 'react';
import { EuiLoadingSpinner, EuiDescriptionList, EuiIcon, EuiCallOut, EuiSpacer, EuiButton } from '@elastic/eui';
import chrome from 'ui/chrome';
import { AppState } from '../../react-services/app-state';
import { PatternHandler } from '../../react-services/pattern-handler';
import { getAngularModule } from 'plugins/kibana/discover/kibana_services';
import { WazuhConfig } from '../../react-services/wazuh-config';
import { GenericRequest } from '../../react-services/generic-request';
import { ApiCheck } from '../../react-services/wz-api-check';
import { ApiRequest } from '../../react-services/api-request';
import { SavedObject } from '../../react-services/saved-objects';
import { ErrorHandler } from '../../react-services/error-handler';
import { toastNotifications } from 'ui/notify';

export class HealthCheck extends Component {
    constructor(props) {
        super(props);
        this.state = {
            checks: [],
            results: [],
            errors: []
        };
    }
    async componentDidMount() {
        const app = getAngularModule('app/wazuh');
        this.$rootScope = app.$injector.get('$rootScope');
        this.load();
    }

    showToast = (color, title, text, time) => {
        toastNotifications.add({
            color: color,
            title: title,
            text: text,
            toastLifeTimeMs: time
        });
    };

    /**
     * Manage an error
     */
    handleError(error) {
        this.setState({
            errors: this.state.errors.push(
                ErrorHandler.handle(error, 'Health Check', { silent: true })
            )
        });
    }

    /**
     * This validates a pattern
     */
    async checkPatterns() {
        try {
            const patternId = AppState.getCurrentPattern();
            let patternTitle = '';
            let results = this.state.results;
            let errors = this.state.errors;
            if (this.state.checks.pattern) {
                const i = this.state.results.map(item => item.id).indexOf(2);
                let patternData = patternId ? await SavedObject.existsIndexPattern(patternId) : false;
                if (!patternData) patternData = {};
                patternTitle = patternData.title;
                /* This extra check will work as long as Wazuh monitoring index ID is wazuh-monitoring-3.x-*.
                   Currently is not possible to change that index pattern as it has always been created on our backend.
                   This extra check checks if the index pattern exists for the current logged in user
                   in case it doesn't exist, the index pattern is automatically created. This is necessary to make it work with Opendistro multinenancy
                   as every index pattern is stored in its current tenant .kibana-tenant-XX index. 
                   */
                try {
                    await SavedObject.existsMonitoringIndexPattern('wazuh-monitoring-3.x-*'); //this checks if it exists, if not it automatically creates the index pattern
                } catch (err) { }
                if (!patternData.status) {
                    const patternList = await PatternHandler.getPatternList("healthcheck");
                    if (patternList.length) {
                        const currentPattern = patternList[0].id;
                        AppState.setCurrentPattern(currentPattern);
                        return this.checkPatterns();
                    } else {
                        errors.push('The selected index-pattern is not present.');
                        results[i].description = <span><EuiIcon type="alert" color="danger" ></EuiIcon> Error</span>;
                    }
                } else {
                    results[i].description = <span><EuiIcon type="check" color="secondary" ></EuiIcon> Ready</span>;
                }
                this.setState({ results, errors });
            }

            if (this.state.checks.template) {
                if (!patternTitle) {
                    var patternData = await SavedObject.existsIndexPattern(patternId);
                    patternTitle = patternData.title;
                }
                const i = results.map(item => item.id).indexOf(3);
                const templateData = await GenericRequest.request(
                    'GET',
                    `/elastic/template/${patternTitle}`
                );
                if (!templateData.data.status) {
                    errors.push('No template found for the selected index-pattern.');
                    results[i].description = <span><EuiIcon type="alert" color="danger" ></EuiIcon> Error</span>;
                } else {
                    results[i].description = <span><EuiIcon type="check" color="secondary" ></EuiIcon> Ready</span>;
                }
                this.setState({ results, errors });
            }
            return;
        } catch (error) {
            this.handleError(error);
        }
    }

    async trySetDefault() {
        try {
            const response = await GenericRequest.request('GET', '/hosts/apis');
            const hosts = response.data;
            const errors = [];

            if (hosts.length) {
                for (var i = 0; i < hosts.length; i++) {
                    try {
                        const API = await ApiCheck.checkApi(hosts[i]);
                        if (API && API.data) {
                            return hosts[i].id;
                        }
                    } catch (err) {
                        errors.push(`Could not connect to API with id: ${hosts[i].id}: ${err.message || err}`);
                    }
                }
                if (errors.length) {
                    let err = this.state.errors;
                    errors.forEach(error => err.push(error));
                    this.setState({ errors: err })
                    return Promise.reject('No API available to connect.');
                }
            }
        } catch (err) {
            return Promise.reject(`Error connecting to API: ${err}`);
        }
    }

    /**
     * This attempts to connect with API
     */
    async checkApiConnection() {
        let results = this.state.results;
        let errors = this.state.errors;
        try {
            const currentApi = JSON.parse(AppState.getCurrentAPI() || '{}');
            if (this.state.checks.api && currentApi && currentApi.id) {
                let data;
                try {
                    data = await ApiCheck.checkStored(currentApi.id);
                } catch (err) {
                    try {
                        const newApi = await this.trySetDefault();
                        data = await ApiCheck.checkStored(newApi, true);
                    } catch (err2) {
                        throw err2
                    };
                }

                if (((data || {}).data || {}).idChanged) {
                    this.showToast(
                        'warning',
                        'Selected Wazuh API has been updated',
                        '',
                        3000
                    );
                    const apiRaw = JSON.parse(AppState.getCurrentAPI());
                    AppState.setCurrentAPI(
                        JSON.stringify({ name: apiRaw.name, id: data.data.idChanged })
                    );
                }
                //update cluster info
                const cluster_info = (((data || {}).data || {}).data || {})
                    .cluster_info;
                if (cluster_info) {
                    AppState.setClusterInfo(cluster_info);
                }
                const i = results.map(item => item.id).indexOf(0);
                if (data === 3099) {
                    errors.push('Wazuh not ready yet.');
                    results[i].description = <span><EuiIcon type="alert" color="danger" ></EuiIcon> Error</span>;
                    if (this.checks.setup) {
                        const i = results.map(item => item.id).indexOf(1);
                        results[i].description = <span><EuiIcon type="alert" color="danger" ></EuiIcon> Error</span>;
                    }
                    this.setState({ results, errors });
                } else if (data.data.error || data.data.data.apiIsDown) {
                    errors.push(data.data.data.apiIsDown ? 'Wazuh API is down.' : `Error connecting to the API.${data.data.error && data.data.error.message ? ` ${data.data.error.message}` : ''}`);
                    results[i].description = <span><EuiIcon type="alert" color="danger" ></EuiIcon> Error</span>;
                    results[i + 1].description = <span><EuiIcon type="alert" color="danger" ></EuiIcon> Error</span>;
                    this.setState({ results, errors });
                } else {
                    results[i].description = <span><EuiIcon type="check" color="secondary" ></EuiIcon> Ready</span>;
                    if (this.state.checks.setup) {
                        const versionData = await ApiRequest.request(
                            'GET',
                            '/version',
                            {}
                        );
                        const apiVersion = versionData.data.data;
                        const setupData = await GenericRequest.request(
                            'GET',
                            '/api/setup'
                        );
                        if (!setupData.data.data['app-version']) {
                            errors.push('Error fetching app version');
                        };
                        if (!apiVersion) {
                            errors.push('Error fetching Wazuh API version');
                        };
                        this.setState({ results, errors });
                        const apiSplit = apiVersion.split('v')[1].split('.');
                        const appSplit = setupData.data.data['app-version'].split('.');

                        const i = results.map(item => item.id).indexOf(1);
                        if (apiSplit[0] !== appSplit[0] || apiSplit[1] !== appSplit[1]) {
                            errors.push(
                                'API version mismatch. Expected v' +
                                setupData.data.data['app-version']
                            );
                            results[i].description = <span><EuiIcon type="alert" color="danger" ></EuiIcon> Error</span>;
                            this.setState({ results, errors });
                        } else {
                            results[i].description = <span><EuiIcon type="check" color="secondary" ></EuiIcon> Ready</span>;
                            this.setState({ results });
                        }
                    }
                }
            }
            return;
        } catch (error) {
            results[0].description = <span><EuiIcon type="alert" color="danger" ></EuiIcon> Error</span>;
            results[1].description = <span><EuiIcon type="alert" color="danger" ></EuiIcon> Error</span>;
            this.setState({ results });
            AppState.removeNavigation();
            if (error && error.data && error.data.code && error.data.code === 3002) {
                return error;
            } else {
                this.handleError(error);
            }
        }
    }

    /**
     * On controller loads
     */
    async load() {
        try {
            const wazuhConfig = new WazuhConfig();
            const configuration = wazuhConfig.getConfig();
            AppState.setPatternSelector(configuration['ip.selector']);
            let checks = {};
            checks.pattern = configuration['checks.pattern'];
            checks.template = configuration['checks.template'];
            checks.api = configuration['checks.api'];
            checks.setup = configuration['checks.setup'];
            checks.fields = configuration['checks.fields'];
            const results = []
            results.push(
                {
                    id: 0,
                    title: 'Check Wazuh API connection',
                    description: checks.api ? <span><EuiLoadingSpinner size="m" /> Checking...</span> : 'Disabled'
                },
                {
                    id: 1,
                    title: 'Check for Wazuh API version',
                    description: checks.setup ? <span><EuiLoadingSpinner size="m" /> Checking...</span> : 'Disabled'
                },
                {
                    id: 2,
                    title: 'Check Elasticsearch index pattern',
                    description: checks.pattern ? <span><EuiLoadingSpinner size="m" /> Checking...</span> : 'Disabled'
                },
                {
                    id: 3,
                    title: 'Check Elasticsearch template',
                    description: checks.template ? <span><EuiLoadingSpinner size="m" /> Checking...</span> : 'Disabled'
                },
                {
                    id: 4,
                    title: 'Check index pattern known fields',
                    description: checks.fields ? <span><EuiLoadingSpinner size="m" /> Checking...</span> : 'Disabled'
                }
            );
            this.setState({ checks, results },
                async () => {
                    await Promise.all([this.checkPatterns(), this.checkApiConnection()]);
                    if (checks.fields) {
                        const i = results.map(item => item.id).indexOf(4);
                        try {
                            await GenericRequest.request('GET', '/elastic/known-fields/all', {});
                            results[i].description = <span><EuiIcon type="check" color="secondary" ></EuiIcon> Ready</span>;
                            this.setState({ results });
                        } catch (error) {
                            results[i].description = <span><EuiIcon type="alert" color="danger" ></EuiIcon> Error</span>;
                            this.setState({ results }, () => this.handleError(error));
                        }
                    }

                    if (!this.state.errors || !this.state.errors.length) {
                        setTimeout(() => {
                            const params = this.$rootScope.previousParams || {};
                            var queryString = Object.keys(params).map(key => key + '=' + params[key]).join('&');
                            const url = 'wazuh#' + (this.$rootScope.previousLocation || '') + '?' + queryString;
                            window.location.assign(
                                chrome.addBasePath(url)
                            );
                            return;
                        }, 300);
                    }
                    return;
                });
        } catch (error) {
            this.handleError(error);
        }
    }

    goApp() {
        window.location.href = '/app/wazuh#/settings';
    }

    render() {
        const logo_url = chrome.addBasePath('/plugins/wazuh/img/icon_blue.svg');
        return (
            <div className="health-check">
                <EuiLoadingSpinner className="health-check-loader" />
                <img src={logo_url} className="health-check-logo" alt=""></img>
                <div className="margin-top-30">
                    <EuiDescriptionList
                        textStyle="reverse"
                        align="center"
                        type="column"
                        listItems={this.state.results}
                    />
                </div>
                <EuiSpacer size="xxl" />
                {this.state.errors.map(error => (
                    <EuiCallOut title={error} color="danger" iconType="alert"></EuiCallOut>)
                )}
                <EuiSpacer size="xxl" />
                {this.state.errors.length && (
                    <EuiButton
                        fill
                        onClick={() => this.goApp()}>
                        Go to App
                    </EuiButton>
                )}
            </div >
        );
    }
};
