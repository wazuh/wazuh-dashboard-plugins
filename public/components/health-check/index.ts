export { HealthCheck } from './container/health-check.container';
