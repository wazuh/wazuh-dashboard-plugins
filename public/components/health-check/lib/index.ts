export { checkKibanaSettings } from './check-kibana-settings';
export { checkKibanaSettingsTimeFilter } from './check-time-filter';
export { checkKibanaSettingsMaxBuckets } from './check-max-buckets';
