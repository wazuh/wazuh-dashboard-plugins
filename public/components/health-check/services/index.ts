/*
 * Wazuh app - Health check services
 *
 * Copyright (C) 2015-2021 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 *
 */

export * from './check-api.service';
export * from './check-fields.service';
export * from './check-kibana-settings';
export * from './check-max-buckets';
export * from './check-pattern.service';
export * from './check-pattern-support.service';
export * from './check-template.service';
export * from './check-time-filter';
export * from './check-setup.service';
