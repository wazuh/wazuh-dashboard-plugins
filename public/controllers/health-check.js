/*
 * Wazuh app - Heakthcheck controller
 * Copyright (C) 2018 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 */
import { uiModules } from 'ui/modules'

const app = uiModules.get('app/wazuh', []);

app.controller('healthCheck', 
function ($scope, $rootScope, $timeout, $location, 
          courier, genericReq, apiReq, appState, testAPI, 
          errorHandler, wazuhConfig
) {
    const checks = {
        api     : true,
        pattern : true,
        setup   : true,
        template: true
    };

    $scope.errors          = [];
    $scope.processedChecks = 0;
    $scope.totalChecks     = 0;

    const handleError = error => {
        errorHandler.handle(error,'Health Check');
        $scope.errors.push(errorHandler.handle(error,'Health Check',false,true));
    };

    const checkPatterns = async () => {
        try {
            const data         = await courier.indexPatterns.get(appState.getCurrentPattern());
            const patternTitle = data.title;

            if(checks.pattern) {
                const i = $scope.results.map(item => item.id).indexOf(2);
                const patternData = await genericReq.request('GET', `/api/wazuh-elastic/pattern/${patternTitle}`);
                if (!patternData.data.status) {
                    $scope.errors.push("The selected index-pattern is not present.");
                    $scope.results[i].status = 'Error';
                } else {
                    $scope.processedChecks++;
                    $scope.results[i].status = 'Ready';
                }
            }

            if(checks.template) {
                const i = $scope.results.map(item => item.id).indexOf(3);
                const templateData = await genericReq.request('GET', `/api/wazuh-elastic/template/${patternTitle}`);
                if (!templateData.data.status) {
                    $scope.errors.push("No template found for the selected index-pattern.");
                    $scope.results[i].status = 'Error';
                } else {
                    $scope.processedChecks++;
                    $scope.results[i].status = 'Ready';
                }
            }

            return;
        } catch (error) {
            handleError(error);
        }
    };

    const checkApiConnection = async () => {
        try {
            if(checks.api) {
                const data = await testAPI.check_stored(JSON.parse(appState.getCurrentAPI()).id);

                const i = $scope.results.map(item => item.id).indexOf(0);
                if (data.data.error || data.data.data.apiIsDown) {
                    $scope.errors.push("Error connecting to the API.");
                    $scope.results[i].status = 'Error';
                } else {
                    $scope.processedChecks++;
                    $scope.results[i].status = 'Ready';
                    if(checks.setup) {
                        const versionData = await apiReq.request('GET', '/version', {});
                        const apiVersion  = versionData.data.data;
                        const setupData   = await genericReq.request('GET', '/api/wazuh-elastic/setup');
                        if(!setupData.data.data["app-version"] || !apiVersion){
                            errorHandler.handle('Error fetching app version or API version','Health Check');
                            $scope.errors.push('Error fetching version');
                        }
                        const apiSplit = apiVersion.split('v')[1].split('.');
                        const appSplit = setupData.data.data["app-version"].split('.');

                        const i = $scope.results.map(item => item.id).indexOf(1);
                        if (apiSplit[0] !== appSplit[0] || apiSplit[1] !== appSplit[1]) {
                            $scope.errors.push("API version mismatch. Expected v" + setupData.data.data["app-version"]);
                            $scope.results[i].status = 'Error';
                        } else {
                            $scope.processedChecks++;                            
                            $scope.results[i].status = 'Ready';
                        }
                    }
                }
            } else {
                if(checks.setup) $scope.processedChecks++;
            }

            return;
        } catch(error) {
            handleError(error);
        }
    };

    $scope.goApp = () => $location.path($rootScope.previousLocation || '/');
    
    $scope.results = [];

    const load = async () => {
        try {
            const configuration = wazuhConfig.getConfig();
            
            appState.setPatternSelector(configuration['ip.selector']);
            
            checks.pattern  = configuration['checks.pattern'];
            checks.template = configuration['checks.template'];
            checks.api      = configuration['checks.api'];
            checks.setup    = configuration['checks.setup'];

            $scope.results.push({ id:0,description: 'Check Wazuh API connection',status: checks.api ? 'Checking...' : 'disabled' });
            $scope.results.push({ id:1,description: 'Check for Wazuh API version',status: checks.setup ? 'Checking...' : 'disabled' });
            $scope.results.push({ id:2,description: 'Check Elasticsearch index pattern',status: checks.pattern ? 'Checking...' : 'disabled' });
            $scope.results.push({ id:3,description: 'Check Elasticsearch template',status: checks.template ? 'Checking...' : 'disabled'});

            for(let key in checks) $scope.totalChecks += (checks[key]) ? 1 : 0;

            if ($scope.totalChecks == 0) $scope.zeroChecks = true;

            await Promise.all([ checkPatterns(), checkApiConnection() ]);

            $scope.checksDone = true;
            if(!$scope.errors || !$scope.errors.length) {
                $timeout(() => $location.path($rootScope.previousLocation || '/'), 800);
                return;
            }

            if(!$scope.$$phase) $scope.$digest();
            return;
        } catch (error) {
            errorHandler.handle(error,'Health Check');
        }
    }

    load();

});
