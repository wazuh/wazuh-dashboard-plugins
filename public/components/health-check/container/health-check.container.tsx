/*
 * Wazuh app - Health Check Component
 *
 * Copyright (C) 2015-2022 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 *
 */

import {
  EuiButton,
  EuiCallOut,
  EuiDescriptionList,
  EuiSpacer,
} from '@elastic/eui';
import React, { Fragment, useState, useEffect, useRef } from 'react';
import {
  EuiFlexGroup,
  EuiFlexItem
} from '@elastic/eui';
import { AppState, ErrorHandler } from '../../../react-services';
import { useAppConfig, useRootScope } from '../../../components/common/hooks';
import {
  checkApiService,
  checkPluginPlatformSettings,
  checkIndexPatternService,
  checkPatternSupportService,
  checkSetupService,
} from '../services';
import { CheckResult } from '../components/check-result';
import { withErrorBoundary, withReduxProvider } from '../../common/hocs';
import { getHttp } from '../../../kibana-services';
import {
  HEALTH_CHECK_REDIRECTION_TIME,
  PLUGIN_PLATFORM_SETTING_NAME_MAX_BUCKETS,
  PLUGIN_PLATFORM_SETTING_NAME_METAFIELDS,
  PLUGIN_PLATFORM_SETTING_NAME_TIME_FILTER,
  WAZUH_INDEX_TYPE_MONITORING,
  WAZUH_INDEX_TYPE_STATISTICS,
  WAZUH_PLUGIN_PLATFORM_SETTING_MAX_BUCKETS,
  WAZUH_PLUGIN_PLATFORM_SETTING_METAFIELDS,
  WAZUH_PLUGIN_PLATFORM_SETTING_TIME_FILTER,
} from '../../../../common/constants';

import { getDataPlugin } from '../../../kibana-services';
import { CheckLogger } from '../types/check_logger';
import { compose } from 'redux';
import { getThemeAssetURL, getAssetURL } from '../../../utils/assets';

const checks = {
  api: {    
    title: 'Check Wazuh API connection',
    label: 'API connection',
    validator: checkApiService,
    awaitFor: [],
    canRetry: true,
  },
  setup: {
    title: 'Check Wazuh API version',
    label: 'API version',
    validator: checkSetupService,
    awaitFor: ["api"],
  },
  pattern: {
    title: 'Check alerts index pattern',
    label: 'Alerts index pattern',
    validator: checkIndexPatternService,
    awaitFor: [],
    shouldCheck: true,
    canRetry: true,
  },
  patternMonitoring: {
    title: 'Check monitoring index pattern',
    label: 'Monitoring index pattern',
    validator: (appConfig) => checkPatternSupportService(appConfig.data['wazuh.monitoring.pattern'], WAZUH_INDEX_TYPE_MONITORING),
    awaitFor: [],
    shouldCheck: true,
    canRetry: true,
  },
  patternStatistics: {
    title: 'Check statistics index pattern',
    label: 'Statistics index pattern',
    validator: (appConfig) => checkPatternSupportService(`${appConfig.data['cron.prefix']}-${appConfig.data['cron.statistics.index.name']}-*`, WAZUH_INDEX_TYPE_STATISTICS),
    awaitFor: [],
    shouldCheck: true,
    canRetry: true,
  },
  maxBuckets: {
    title: `Check ${PLUGIN_PLATFORM_SETTING_NAME_MAX_BUCKETS} setting`,
    label: `${PLUGIN_PLATFORM_SETTING_NAME_MAX_BUCKETS} setting`,
    validator: checkPluginPlatformSettings(PLUGIN_PLATFORM_SETTING_NAME_MAX_BUCKETS, WAZUH_PLUGIN_PLATFORM_SETTING_MAX_BUCKETS),
    awaitFor: [],
    canRetry: true,
  },
  metaFields: {
    title: `Check ${PLUGIN_PLATFORM_SETTING_NAME_METAFIELDS} setting`,
    label: `${PLUGIN_PLATFORM_SETTING_NAME_METAFIELDS} setting`,
    validator: checkPluginPlatformSettings(PLUGIN_PLATFORM_SETTING_NAME_METAFIELDS, WAZUH_PLUGIN_PLATFORM_SETTING_METAFIELDS),
    awaitFor: [],
    canRetry: true,
  },
  timeFilter: {
    title: `Check ${PLUGIN_PLATFORM_SETTING_NAME_TIME_FILTER} setting`,
    label: `${PLUGIN_PLATFORM_SETTING_NAME_TIME_FILTER} setting`,
    validator: checkPluginPlatformSettings(PLUGIN_PLATFORM_SETTING_NAME_TIME_FILTER, JSON.stringify(WAZUH_PLUGIN_PLATFORM_SETTING_TIME_FILTER), (checkLogger: CheckLogger, options: {defaultAppValue: any}) => {
      getDataPlugin().query.timefilter.timefilter.setTime(WAZUH_PLUGIN_PLATFORM_SETTING_TIME_FILTER)
        && checkLogger.action(`Timefilter set to ${JSON.stringify(options.defaultAppValue)}`);
    }),
    awaitFor: [],
    canRetry: true,
  }
};

function HealthCheckComponent() {
  const [checkErrors, setCheckErrors] = useState<{[key:string]: []}>({});
  const [checksReady, setChecksReady] = useState<{[key: string]: boolean}>({});
  const [isDebugMode, setIsDebugMode] = useState<boolean>(false);
  const appConfig = useAppConfig();
  const checksInitiated = useRef(false);
  const $rootScope = useRootScope();

  const redirectionPassHealthcheck = () => {
    const params = $rootScope.previousParams || {};
    const queryString = Object.keys(params).map(key => key + '=' + params[key]).join('&');
    const url = '/app/wazuh#' + ($rootScope.previousLocation || '') + '?' + queryString;
    window.location.href = getHttp().basePath.prepend(url);
  };

  useEffect(() => {
    if (appConfig.isReady && !checksInitiated.current) {
      checksInitiated.current = true;
      AppState.setPatternSelector(appConfig.data['ip.selector']);
    }
  }, [appConfig]);

  useEffect(() => {
    // Redirect to app when all checks are ready
    Object.keys(checks)
      .every(check => checksReady[check])
    && !isDebugMode && (() => setTimeout(redirectionPassHealthcheck, HEALTH_CHECK_REDIRECTION_TIME)
      )()
  }, [checksReady]);

  useEffect(() => {
    // Check if Health should not redirect automatically (Debug mode)
    setIsDebugMode(window.location.href.includes('debug'));
  },[]);

  const handleErrors = (checkID, errors, parsed) => {
    const newErrors = parsed
      ? errors.map((error) =>
          ErrorHandler.handle(error, 'Health Check', { warning: false, silent: true })
        )
      : errors;
    setCheckErrors((prev) => ({...prev, [checkID]: newErrors}));
  };

  const cleanErrors = (checkID: string) => {
    delete checkErrors[checkID];
    setCheckErrors({...checkErrors});
  }

  const handleCheckReady = (checkID, isReady) => {    
    setChecksReady(prev =>  ({...prev, [checkID]: isReady}));
  }


  const logoUrl = getHttp().basePath.prepend(appConfig.data['customization.logo.healthcheck'] ? getAssetURL(appConfig.data['customization.logo.healthcheck']) : getThemeAssetURL('logo.svg'));
  const thereAreErrors = Object.keys(checkErrors).length > 0;

  const renderChecks = () => {
    const showLogButton = (thereAreErrors || isDebugMode);
    return Object.keys(checks).map((check, index) => {
      return (
        <CheckResult
          showLogButton={showLogButton}
          key={`health_check_check_${check}`}
          name={check}
          title={checks[check].title}
          awaitFor={checks[check].awaitFor}
          shouldCheck={checks[check].shouldCheck || appConfig.data[`checks.${check}`]}
          validationService={checks[check].validator(appConfig)}
          handleErrors={handleErrors}
          cleanErrors={cleanErrors}
          isLoading={appConfig.isLoading}
          handleCheckReady= {handleCheckReady}
          checksReady={checksReady}
          canRetry={checks[check].canRetry}
        />
      );
    });
  };

  const renderErrors = () => {
    return Object.keys(checkErrors).map((checkID) => 
      checkErrors[checkID].map((error, index) => (
        <Fragment key={index}>
          <EuiCallOut
            title={(<>{`[${checks[checkID].label}]`} <span dangerouslySetInnerHTML={{__html: error}}></span></>)}
            color="danger"
            iconType="alert"
            style={{ textAlign: 'left' }}
            data-test-subj='callOutError'
          ></EuiCallOut>
          <EuiSpacer size="xs" />
        </Fragment>
      ))
    ) 
  };

  return (
    <div className="health-check">
      <img src={logoUrl} className="health-check-logo" alt=""></img>
      <div className="margin-top-30">
        <EuiDescriptionList textStyle="reverse" align="center" type="column">
          {renderChecks()}
        </EuiDescriptionList>
      </div>
      {thereAreErrors && (
        <>
          <EuiSpacer size="xl" />
          {renderErrors()}
        </>
      )}
      {(thereAreErrors || isDebugMode) && (
        <>
          <EuiSpacer size="xl" />
          <EuiFlexGroup justifyContent='center'>
            {thereAreErrors && (
              <EuiFlexItem grow={false}>
                <EuiButton fill href={getHttp().basePath.prepend('/app/wazuh#/settings')}>
                  Go to Settings
                </EuiButton>
              </EuiFlexItem>
            )}
            {isDebugMode && Object.keys(checks).every(check => checksReady[check]) && (
              <EuiFlexItem grow={false}>
                <EuiButton fill onClick={redirectionPassHealthcheck}>
                  Continue
                </EuiButton>
              </EuiFlexItem>
            )}
          </EuiFlexGroup>
        </>
      )
      }
      <EuiSpacer size="xl" />
    </div>
  );
}

export const HealthCheck = compose (withErrorBoundary,withReduxProvider) (HealthCheckComponent);

export const HealthCheckTest = HealthCheckComponent;


