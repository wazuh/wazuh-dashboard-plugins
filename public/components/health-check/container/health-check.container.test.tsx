/*
 * Wazuh app - Health Check Component - Test
 *
 * Copyright (C) 2015-2021 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 *
 */

import React from 'react';
import { mount, shallow } from 'enzyme';
import { HealthCheckTest } from './health-check.container';


jest.mock('../../../components/common/hooks/use-app-config', () => ({
  useAppConfig: () => ({
    isReady: true,
    isLoading: false,
    data: {
      'ip.selector': 1,
      'checks.metaFields': true,
      'checks.timerFilter': true,
      'checks.api': true,
      'checks.setup': true,
      'checks.pattern': true,
      'checks.template': true,
      'checks.fields': true,
    },
  }),
}));

jest.mock('../services', () => ({
  checkPatternService: () => ({ errors: [] }),
  checkTemplateService: () => ({ errors: [] }),
  checkApiService: () => ({ errors: [] }),
  checkSetupService: () => ({ errors: [] }),
  checkFieldsService: () => ({ errors: [] }),
  checkKibanaSettings: () => ({ errors: [] }),
  checkKibanaSettingsTimeFilter: () => ({ errors: [] }),
  checkPatternSupportService: () => ({ errors: [] })
}));

jest.mock('../components/check-result', () => ({
  CheckResult: () => () => <></>,
}));

jest.mock('../../../react-services', () => ({
  AppState: {
    setPatternSelector: () => {},
  },
  ErrorHandler: {
    handle: (error) => error
  }
}));

jest.mock('../../../kibana-services', () => ({
  getHttp: () => ({
    basePath: {
      prepend: (str) => str
    }
  }),
}));

describe('Health Check container', () => {
  test('should render a Health check screen', () => {
    const component = shallow(<HealthCheckTest />);

    expect(component).toMatchSnapshot();
  });

  test('should render a Health check screen with error', () => {
    const component = mount(<HealthCheckTest />);

    component.find('CheckResult').at(1).invoke('handleErrors')('setup',['Test error']); // invoke is wrapped with act to await for setState

    const callOutError = component.find('EuiCallOut');
    expect(callOutError.text()).toBe('Test error');
  });
});
