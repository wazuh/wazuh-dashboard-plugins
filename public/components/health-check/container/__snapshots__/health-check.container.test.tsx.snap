// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Health Check container should render a Health check screen 1`] = `
<div
  className="health-check"
>
  <img
    alt=""
    className="health-check-logo"
    src="/plugins/wazuh/assets/images/themes/light/logo.svg"
  />
  <div
    className="margin-top-30"
  >
    <EuiDescriptionList
      align="center"
      textStyle="reverse"
      type="column"
    >
      <CheckResult
        awaitFor={Array []}
        canRetry={true}
        checksReady={Object {}}
        cleanErrors={[Function]}
        handleCheckReady={[Function]}
        handleErrors={[Function]}
        isLoading={false}
        key="health_check_check_api"
        name="api"
        shouldCheck={true}
        showLogButton={false}
        title="Check Wazuh API connection"
        validationService={[Function]}
      />
      <CheckResult
        awaitFor={
          Array [
            "api",
          ]
        }
        checksReady={Object {}}
        cleanErrors={[Function]}
        handleCheckReady={[Function]}
        handleErrors={[Function]}
        isLoading={false}
        key="health_check_check_setup"
        name="setup"
        shouldCheck={true}
        showLogButton={false}
        title="Check Wazuh API version"
        validationService={[Function]}
      />
      <CheckResult
        awaitFor={Array []}
        canRetry={true}
        checksReady={Object {}}
        cleanErrors={[Function]}
        handleCheckReady={[Function]}
        handleErrors={[Function]}
        isLoading={false}
        key="health_check_check_pattern"
        name="pattern"
        shouldCheck={true}
        showLogButton={false}
        title="Check alerts index pattern"
        validationService={[Function]}
      />
      <CheckResult
        awaitFor={Array []}
        canRetry={true}
        checksReady={Object {}}
        cleanErrors={[Function]}
        handleCheckReady={[Function]}
        handleErrors={[Function]}
        isLoading={false}
        key="health_check_check_patternMonitoring"
        name="patternMonitoring"
        shouldCheck={true}
        showLogButton={false}
        title="Check monitoring index pattern"
        validationService={[Function]}
      />
      <CheckResult
        awaitFor={Array []}
        canRetry={true}
        checksReady={Object {}}
        cleanErrors={[Function]}
        handleCheckReady={[Function]}
        handleErrors={[Function]}
        isLoading={false}
        key="health_check_check_patternStatistics"
        name="patternStatistics"
        shouldCheck={true}
        showLogButton={false}
        title="Check statistics index pattern"
        validationService={[Function]}
      />
      <CheckResult
        awaitFor={Array []}
        canRetry={true}
        checksReady={Object {}}
        cleanErrors={[Function]}
        handleCheckReady={[Function]}
        handleErrors={[Function]}
        isLoading={false}
        key="health_check_check_maxBuckets"
        name="maxBuckets"
        shouldCheck={true}
        showLogButton={false}
        title="Check timelion:max_buckets setting"
      />
      <CheckResult
        awaitFor={Array []}
        canRetry={true}
        checksReady={Object {}}
        cleanErrors={[Function]}
        handleCheckReady={[Function]}
        handleErrors={[Function]}
        isLoading={false}
        key="health_check_check_metaFields"
        name="metaFields"
        shouldCheck={true}
        showLogButton={false}
        title="Check metaFields setting"
      />
      <CheckResult
        awaitFor={Array []}
        canRetry={true}
        checksReady={Object {}}
        cleanErrors={[Function]}
        handleCheckReady={[Function]}
        handleErrors={[Function]}
        isLoading={false}
        key="health_check_check_timeFilter"
        name="timeFilter"
        showLogButton={false}
        title="Check timepicker:timeDefaults setting"
      />
    </EuiDescriptionList>
  </div>
  <EuiSpacer
    size="xl"
  />
</div>
`;
